File: 61.js
// LeetCode 61: Rotate List
// References:https://leetcode.com/problems/rotate-list/solutions/1838161/algo-explained-easy-js-97-fast-commented/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (!head) return head;
    let count = 0, ptr = head;

    // Step 1: count list nodes
    while (ptr) {
        count++;ptr = ptr.next;
    }

    // Step 2: Number of rotations are now restricted within limit
    k = k % count;
    let prev = head;
    ptr = head;

    //Step 3: Moving one pointer k positions ahead
	while (k--) {
		ptr = ptr.next;
	}

    //Step 4: The actual magic, explained above
	while (ptr.next) {
		prev = prev.next;
		ptr = ptr.next;
	}

    //Step 5: Simply modifying the head and last node
	ptr.next = head;
	head = prev.next;
	prev.next = null;
	return head;
};

File: 61.java
// LeetCode 61: Rotate List
// References:https://leetcode.com/problems/rotate-list/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null) return head;

        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy, slow = dummy;

        int i;
        for (i = 0; fast.next != null; i++) // Get the total length
            fast = fast.next;
        
        for (int j = i - k % i; j > 0; j--) // Get the i - k % i th node
            slow = slow.next;

        fast.next = dummy.next; // Do the rotation
        dummy.next = slow.next;
        slow.next = null;

        return dummy.next;
    }
}
